File: 04-linked-lists/04_merge_two_lists.c
#include "../test_runner.h"

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

ListNode* make_node(int val) {
    ListNode* n = malloc(sizeof(ListNode));
    n->val = val; n->next = NULL;
    return n;
}

ListNode* make_list(int* arr, int n) {
    if (n == 0) return NULL;
    ListNode* head = make_node(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; i++) {
        curr->next = make_node(arr[i]);
        curr = curr->next;
    }
    return head;
}

void to_array(ListNode* head, int* out, int* len) {
    *len = 0;
    while (head) { out[(*len)++] = head->val; head = head->next; }
}


ListNode* merge_two_lists(ListNode* l1, ListNode* l2) {
    return NULL;
}

int main() {
    int out[10], len;
    to_array(merge_two_lists(make_list(ARR(1,2,4),3), make_list(ARR(1,3,4),3)), out, &len);
    TEST_ARR("basic", out, ARR(1,1,2,3,4,4), 6);
    to_array(merge_two_lists(NULL, make_list(ARR(0),1)), out, &len);
    TEST_ARR("one empty", out, ARR(0), 1);
    to_array(merge_two_lists(NULL, NULL), out, &len);
    TEST_INT("both empty", len, 0);
    TEST_SUMMARY();
}

File: 04-linked-lists/04_merge_two_lists.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


def make_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def to_array(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr


def merge_two_lists(l1, l2):
    pass


def solve(a1, a2):
    return to_array(merge_two_lists(make_list(a1), make_list(a2)))


run_tests(solve, [
    (([1,2,4], [1,3,4]),  [1,1,2,3,4,4]),
    (([], []),             []),
    (([], [0]),            [0]),
    (([1], [2]),           [1,2]),
])
